Mu2eUtilities/src/CaloPulseShape.cc

```diff
@@ -15,89 +16,86 @@
-       if (pulseFile.IsOpen()) pshape = (TH1F*) pulseFile.Get(histName.c_str());
-
-       if (!pshape) throw cet::exception("CATEGORY")<<"CaloPulseShape:: Hitsogram "<<histName.c_str()
-                                                   <<" from file "<<fileName.c_str()<<" does not exist";
-
-       // smooth the histogram, normalize and resmaple with the desired binning
-       int nbins = int((pshape->GetXaxis()->GetXmax()-pshape->GetXaxis()->GetXmin())/digiSampling_*pulseIntegralSteps_);
-       TH1F pulseShape("ps_temp","ps_temp", nbins, pshape->GetXaxis()->GetXmin(), pshape->GetXaxis()->GetXmax());       
-       TSpline3 spline(pshape);
-       for (int i=1;i<pulseShape.GetNbinsX();++i) pulseShape.SetBinContent(i,spline.Eval(pulseShape.GetBinCenter(i)));
-
-       // Integrated waveform over digitization period for each starting point if required
-       if (doIntegral_) 
-       {
-	   TH1F pulseShapeIntegral(pulseShape);
-	   for (int i=1;i<pulseShapeIntegral.GetNbinsX();++i) 
-	      pulseShapeIntegral.SetBinContent(i,pulseShape.Integral(i, std::min(i+pulseIntegralSteps_-1,nbins-1)));
-	   pulseShape = pulseShapeIntegral;
-       }
-
-       // Copy histogram into vector, including time padding      
-       for (int j=1;j<=nbins;++j)
-       {
-	  int ibin = j-pulseIntegralSteps_;
-	  integralVal_.push_back((ibin>0) ? pulseShape.GetBinContent(ibin) : 0.0);
-       }
-
-       // Normalize waveform such that the max vector is at 1
-       double integralValMax = *max_element(integralVal_.begin(),integralVal_.end());
-       for (auto& v : integralVal_) v/=integralValMax;
-       
-       //calculate the number of bins for the digitized waveform
-       nBinShape_      = int(nbins/pulseIntegralSteps_);
+         if (pulseFile.IsOpen()) pshape = (TH1F*) pulseFile.Get(histName.c_str());
+         if (!pshape) throw cet::exception("CATEGORY")<<"CaloPulseShape:: Hitsogram "<<histName.c_str()
+                                                     <<" from file "<<fileName.c_str()<<" does not exist";
+	 pshape->SetDirectory(0);
+       pulseFile.Close();
+
+
+       // Adjust binning to match digitizer sampling period, shift to zero and normalize
+       int nbins = int((pshape->GetXaxis()->GetXmax()-pshape->GetXaxis()->GetXmin())/digiStep_);
+       TH1F pulseShape("ps","ps", nbins, 0.0, pshape->GetXaxis()->GetXmax()-pshape->GetXaxis()->GetXmin());
+       for (int i=1;i<nbins;++i) pulseShape.SetBinContent(i,pshape->Interpolate(pulseShape.GetBinCenter(i)+pshape->GetXaxis()->GetXmin()));
+       pulseShape.Scale(1.0/pulseShape.GetMaximum(),"nosw2");
+
+       // Fit the top of the pulse shape to smooth small wiggles -> speed up fit convergence (wiggles = local minima = huge pain)
+       // adjust the fitted bounds to get a smooth transition between shape and fit
+       double maxBin = pulseShape.GetMaximumBin();
+       pulseShape.Fit("pol6","q0","O0",pulseShape.GetBinCenter(maxBin)-20, pulseShape.GetBinCenter(maxBin)+40);
```

Comment:
where do the numbers 20 and 40 come from?  They should be parameters (if they could ever change) or constexpr if they can't.

Reply:
I'll make them parameters

Reply:
I smoothed the top of the waveform to avoid local minima and facilitate convergence when we extract the signal, and these two numbers were limits to perform the local fit. I replaced the two numbers by an algorithm that determines the boundaries automatically, and declared the only parameter of the procedure as constexpr.